Work-in-progress converting USC to use MirAL

WebDrake opened the pull request:
This PR has been created to allow for some interim review of the changes being made to help convert USC to use MirAL instead of the mirserver API.  See https://github.com/yunit-io/unity-system-compositor/issues/1 for details and https://github.com/yunit-io/unity-system-compositor/issues/1#issuecomment-321248078 in particular for the main action plan.

As much as possible, changes have been made in an incremental manner, slowly separating out the concerns.  The long-term aim here is to convert the `main` function to using `miral::MirRunner` instead of directly initializing a `mir::Server` instance (note, this is still TBD).  In the short term, the aim has been to preserve all the functionality and command-line options of the original USC: decisions about what to keep and what to drop are being left until things are at a point where we are in a position to start thinking about upstreaming components to MirAL, or replacing them with MirAL equivalents.

WebDrake committed:
Add MirAL dependency to cmake setup

WebDrake committed:
Replace USC display configuration with MirAL equivalent

This patch removes the `add_display_configuration_options_to` function
(which makes direct internal use of `mirserver` functionality), using in
its place the similarly-purposed MirAL `display_configuration_options`.

Note that this is not entirely a drop-in replacement, since the MirAL
function not only reproduces the side-by-side `display-config` of the
USC original, but also adds translucency settings, and rather than
returning a `DisplayConfigurationPolicy` instance directly, returns a
`PixelFormatSelector` (a subclass defined in the MirAL source file that
adds support for the alpha translucency settings).

WebDrake committed:
Use get_options directly instead of usc::Server::the_options

It's not entirely obvious why this abstraction is here, but removing it
means that downstream code can use the same calling convention for both
`usc::Server` and `mir::Server` instances.  This should make it easier
to refactor out the various getter methods in `usc::Server` that depend
on access to options.

WebDrake committed:
Use get_options directly in usc::SystemCompositor::run

This patch removes calls to wrapper methods of `usc::Server` that return
values derived from command-line arguments.  This ensures the resulting
calls will remain valid also for a `mir::Server` instance that lacks the
wrapper methods.

WebDrake committed:
Check for --version flag in `main` method instead of SystemCompositor

This should make it easier to transfer from using `usc::Server` to using
`miral::MirRunner`, since it removes a key `SystemCompositor` dependency
on methods of the `Server` class.

When transitioning to `MirRunner` it should be possible to get the state
of the version flag from a callback used with a `CommandLineOption`.

WebDrake committed:
Move usc::Server::the_dm_connection to usc::SystemCompositor internals

This is part of the work of unpicking the display-manager code from the
`usc::Server` class, so that the latter can be replaced with a simple
`mir::Server` instance.

A callback returning the `shared_ptr<SessionSwitcher>` instance is now
passed as a constructor parameter to `usc::SystemCompositor` (note that
it must be a callback and not the instance itself, because the instance
will not be initialized until after the `SystemCompositor` instance is
constructed).  This enables us to replace the `the_dm_connection` method
with direct instantiation of the `dm_connection` field from inside the
server init callback defined inside `usc::SystemCompositor::run`.  The
new code uses `the_session_switcher` directly rather than abstracting it
away behind a replacement `the_dm_message_handler`: the latter method of
`usc::Server` is simply removed.

Once `usc::Server` has been entirely replaced with a plain `mir::Server`
instance, it should be possible to further rework this code to replace
the `get_options()` calls with `CommandLineOption` instances.

WebDrake committed:
Move ownership of SessionSwitcher outside usc::Server

This patch unpicks one of the tighter couplings of `usc::Server` and the
rest of the USC codebase, particularly with `usc::SystemCompositor`.

The `SessionSwitcher` instance forms the main co-dependency between the
`usc::Server` and `usc::SystemCompositor` class, since it is used in the
`the_dm_message_handler` method and thus the `the_dm_connection` method
used inside `usc::SystemCompositor`.  It is however first used in the
`usc::Server` constructor to override the window manager builder, and it
needs the server options to be available in order to construct it, since
it relies on both the `mir::Server` `--file` flag, and the `--spinner`
flag defined in USC.

In order to unpick this co-dependency, the `shared_ptr<SessionSwitcher>`
ownership has been moved to `main()`, where it is declared but not yet
initialized.  A reference to this is passed to the constructor of the
`usc::Server` class, which initializes it from within the window manager
override callback (!).  This is necessary in order to ensure the correct
sequence of events:

  * initializing the `SessionSwitcher` requires the `ExternalSpinner`
    instance to be initialized first

  * initializing the `ExternalSpinner` class requires access to the
    results of the `--file` command-line option, which can only be
    accessed via `mir::Server::get_options` (since it is a built-in
    flag defined internally by `mir::Server`)

  * `mir::Server::get_options` cannot be called until after the call
    to `mir::Server::apply_settings` at the end of the `usc::Server`
    constructor

  * the `apply_settings()` call will apply the window manager builder
    override, and therefore the callback, ensuring that the session
    switcher is initialized before the `usc::Server` constructor exits.

Since the `SessionSwitcher` stores the `shared_ptr<Spinner>` internally,
we are able to drop it as a field of both the `usc::SystemCompositor`
and `usc::Server` classes, since it is not needed to access directly.
This cleans up a bit of unnecessary `CachedPtr`-based lazy evaluation.
We also switch to using `get_options()` directly to access the results
of the `--file` and `--spinner` flags, in order to avoid depending on
custom methods of `usc::Server`.

Following this patch it should be possible to move the display-manager
methods (`the_dm_connection` and `the_dm_message_handler`) out of the
`usc::Server` class and into `usc::SystemCompositor` (which is where
they are actually used).

WebDrake committed:
Remove get_options wrapper methods from usc::Server

These are no longer used anywhere in the codebase and can all be safely
discarded.

Note that much of the direct use of `get_options` should be replaced by
uses of `CommandLineOption` in future as part of the wider MirAL shift.

WebDrake committed:
Make dbus_bus_address a standalone function exposed by server.h

This should allow us to more easily transition various methods of the
`usc::Server` class to `usc::SystemCompositor`.  In the long run it is
probably going to be possible to move the function to be internal to
`system_compositor.cpp`.

WebDrake committed:
Move creation of screen event handler entirely to usc::SystemCompositor

This patch makes explicit the sequence of events required to create the
screen event handler, replacing `usc::Server` methods that abstracted
away the procedures involved.  In consequence we are also able to remove
the `CachedPtr` instances used by the old code, since the laziness they
enabled seems to never actually have been necessary.

This helps us move closer to being able to replace `usc::Server` with a
plain `mir::Server`.

WebDrake committed:
Move creation of the unity display service to usc::SystemCompositor

This patch removes `the_unity_display_service` method from `usc::Server`
(including the underlying `CachedPtr` field), and instead initializes
the `unity_display_service` field directly from within the init callback
created in `usc::SystemCompositor::run`.

It is still necessary for now to use `usc::Server::the_dbus_event_loop`
when creating the unity display service, but this should be possible to
address in a later patch once other consumers of that method have been
moved out of `usc::Server`.  Use of `usc::Server::the_screen` has been
replaced by a direct use of the `screen` class field initialized earlier
in the callback.

WebDrake committed:
Move creation of the unity input service to usc::SystemCompositor

This patch removes `the_unity_input_service` method from `usc::Server`
(including the underlying `CachedPtr` field), and instead initializes
the `unity_input_service` field directly from within the init callback
created in `usc::SystemCompositor::run`.

It is still necessary for now to use `usc::Server::the_dbus_event_loop`
and `usc::Server::the_input_configuration` when creating the unity input
service, but this should be possible to address in later patches once
other consumers of these methods have been moved out of `usc::Server`.

WebDrake committed:
Make dbus_bus_address() an internal function of system_compositor.cpp

Now that all calling sites of this function have been moved over to the
`usc::SystemCompositor` class, we no longer need it to be accessible
from anywhere else.  Note that this might need to change in future if we
wish to break apart the `SystemCompositor` class itself, but for now it
suffices to keep it as an internal implementation detail.

WebDrake committed:
Move creation of dbus service thread to usc::SystemCompositor

This patch removes the `usc::Server::the_dbus_connection_thread` method
and the underlying `CachedPtr` field.  The `dbus_service_thread` field
of `usc::SystemCompositor` is instead simply initialized in place.

The call to `usc::Server::the_dbus_event_loop` should be possible to
remove in a future patch, since all consumers of this method should now
have been moved to `usc::SystemCompositor`.

WebDrake committed:
Move creation of dbus event loop to usc::SystemCompositor

This patch removes the `usc::Server::the_dbus_event_loop` method and the
underlying `CachedPtr` field and instead creates the necessary instance
inside the init callback defined inside `usc::SystemCompositor::run`.

Since all consumers of the `shared_ptr<DBusEventLoop>` are created from
within the same callback, and assigned to fields of `SystemCompositor`,
this instance should stay valid for the lifetime of the latter class,
and hence for as long as needed by the application.

WebDrake committed:
Make usc::Server inherit from public mir::Server

The last remaining custom methods of `usc::Server` both wrap calls to
underlying methods of `mir::Server` itself.  We therefore cannot get rid
of them without downstream functionality having access to `mir::Server`
methods.

WebDrake committed:
Move creation of MirScreen instance to usc::SystemCompositor

With `usc::Server` now inheriting publicly from `mir::Server`, we can
use the methods of the latter class directly from within the callback
defined in `usc::SystemCompositor::run` when initializing the `screen`
field.

Note that an explicit `shared_ptr<MirScreen>` needs to be created first
in order to call the `register_interest` method, since `Screen` itself
does not inherit from `mir::graphics::DisplayConfigurationObserver`.

WebDrake committed:
Remove private session_switcher field of usc::Server

No remaining class methods make use of this field, so we can drop it and
use only the reference passed to the constructor.

WebDrake committed:
Move creation of Mir input configuration to usc::SystemCompositor

Since this is only used inside the init callback created inside the
`usc::SystemCompositor::run` method, and a copy of the shared pointer
will be kept inside the `UnityInputService` instance generated with it,
we do not need to provide this via a method of `usc::Server`, and we can
consequently dispense with the underlying `CachedPtr` field.

This patch leaves `usc::Server` with no custom fields or methods, only a
custom constructor.

WebDrake committed:
Rework usc::SystemCompositor::run as operator()(mir::Server&)

Now that `usc::Server` inherits publicly from `mir::Server`, and earlier
changes have removed all custom methods of `usc::Server`, we can remove
all explicit `usc::Server` dependencies from `usc::SystemCompositor`.

Replacing the `run` method with an `operator()(mir::Server&)` will allow
`usc::SystemCompositor` to be used as input for `MirRunner::run_with`.
In preparation for this, the call to `Server::run` is moved out to the
`main` function.

WebDrake committed:
Clean up some unnecessary imports and aliases from mir_screen.cpp

WebDrake committed:
Use a pre_init CommandLineOption to implement the --version flag

This method of handling `--version` removes some of the custom arguments
handling blocking the transition to using `miral::MirRunner`.

WebDrake commented on src/server_overrides.cpp:
```diff
@@ -113,13 +115,26 @@ struct StubCookieAuthority : public mir::cookie::Authority
 void usc::server_overrides(mir::Server& server)
 {
-    server.wrap_cursor_listener([&server](std::shared_ptr<mir::input::CursorListener> const& default_)
+    bool enable_hardware_cursor{false};
```
This patch introduces a segfault, for reasons that are obvious in retrospect: the `enable_hardware_cursor` variable will be out of scope by the time the callback passed to `wrap_cursor_listener` is invoked.  This patch should therefore be dropped for the time being (we can always come back to it later).

WebDrake commented on src/system_compositor.cpp:
```diff
@@ -99,7 +123,24 @@ void usc::SystemCompositor::run()
+            }
+            else if (server->get_options()->is_set("debug-without-dm"))
+            {
+                dm_connection = std::make_shared<NullDMMessageHandler>(
```
This patch introduces a segfault when running `unity-system-compositor --debug-without-dm`.  Without having looked in depth yet, my guess would be that `session_switcher` is not yet initialized, probably because the init callback is called before the callback provided to `override_the_window_manager_builder`.

Probably the best way to handle this is to provide `usc::Server` and `usc::SystemCompositor` with a callback that lazily initializes `session_switcher`, instead of just providing the `shared_ptr` directly.

WebDrake replied:
Looking at this with `gdb` suggests the problem is for a different reason (which makes sense given that the preceding patch would have displayed the segfault if it was about initialization of the `SessionSwitcher` instance):

```gdb
Thread 1 "unity-system-co" received signal SIGSEGV, Segmentation fault.
0x00000000004cee42 in (anonymous namespace)::NullDMMessageHandler::start() ()
(gdb) bt
#0  0x00000000004cee42 in (anonymous namespace)::NullDMMessageHandler::start() ()
#1  0x00000000004cf6be in usc::SystemCompositor::run()::{lambda()#1}::operator()() const ()
#2  0x00000000004cff01 in std::_Function_handler<void (), usc::SystemCompositor::run()::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#3  0x00007ffff72ea908 in mir::run_mir(mir::ServerConfiguration&, std::function<void (mir::DisplayServer&)>, std::function<void (int)> const&) ()
   from /usr/lib/x86_64-linux-gnu/libmirserver.so.43
#4  0x00007ffff72f6b9c in mir::Server::run() ()
   from /usr/lib/x86_64-linux-gnu/libmirserver.so.43
#5  0x00000000004cfb62 in usc::SystemCompositor::run() ()
#6  0x00000000004b1bad in main ()
```

WebDrake replied:
More detailed backtrace from a debug build:

```gdb
Thread 1 "unity-system-co" received signal SIGSEGV, Segmentation fault.
0x00000000004cee42 in (anonymous namespace)::NullDMMessageHandler::start (
    this=0xd18e90)
    at /home/joseph/code/yunit-io/unity-system-compositor/src/system_compositor.cpp:90
90	        dm_message_handler->set_active_session(client_name);
(gdb) bt
#0  0x00000000004cee42 in (anonymous namespace)::NullDMMessageHandler::start (
    this=0xd18e90)
    at /home/joseph/code/yunit-io/unity-system-compositor/src/system_compositor.cpp:90
#1  0x00000000004cf6be in usc::SystemCompositor::<lambda()>::operator()(void) const (
    __closure=0x755940)
    at /home/joseph/code/yunit-io/unity-system-compositor/src/system_compositor.cpp:155
#2  0x00000000004cff01 in std::_Function_handler<void(), usc::SystemCompositor::run()::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...)
    at /usr/include/c++/5/functional:1871
#3  0x00007ffff72ea908 in mir::run_mir(mir::ServerConfiguration&, std::function<void (mir::DisplayServer&)>, std::function<void (int)> const&) ()
   from /usr/lib/x86_64-linux-gnu/libmirserver.so.43
#4  0x00007ffff72f6b9c in mir::Server::run() ()
   from /usr/lib/x86_64-linux-gnu/libmirserver.so.43
#5  0x00000000004cfb62 in usc::SystemCompositor::run (this=0x7fffffffdc10)
    at /home/joseph/code/yunit-io/unity-system-compositor/src/system_compositor.cpp:167
#6  0x00000000004b1bad in main (argc=4, argv=0x7fffffffdd98)
    at /home/joseph/code/yunit-io/unity-system-compositor/src/main.cpp:41
```

WebDrake replied:
... so, in fact, it is that `session_switcher` is not initialized when the `NullDMMessageHandler` instance is created.  Rather than the order of the callbacks, my suspicion is that `session_switcher` is not initialized when the `SystemCompositor` constructor is called.

This suggests that the trick here is to provide `SystemCompositor` with a `the_session_switcher` delegate, but to keep `Server` as it is.  Should be straightforward enough in the circumstances.

WebDrake commented:
Updated to fix the segfaults: the `Implement --enable-hardware-cursor flag using miral::CommandLineOption` patch has simply been dropped, while `Move usc::Server::the_dm_connection to usc::SystemCompositor internals` has been reworked to pass a callback to the `SystemCompositor` constructor rather than a `shared_ptr<SessionSwitcher>`.

I've also done a bit of cleanup of some later patches affected by these changes.

WebDrake committed:
Entirely replace usc::Server with mir::Server

Now that usc::Server has no custom methods or fields, it is possible to
entirely strip out that class and use `mir::Server` instead.  All of the
work done inside the old `usc::Server` constructor is now done instead
in the `main()` function.

This ought to leave us with a clearer sequence of events to unpick as we
transition towards using `miral::MirRunner`.

Note that all lambda closures specifying their capture rules as `[this]`
inside `usc::Server` have been updated to use `[&]` instead.  Obviously
we could use more specific captures, but this should make it simpler to
update code in future, and in particular to avoid hassle over capturing
by reference versus by value (which may have an impact for example when
we want to use values set by `CommandLineOption` callbacks).

WebDrake commented:
Updated again to remove the unnecessary separation out of code into `server_overrides()`.  The code now remains in-place (initially in `usc::Server::Server` before being moved out to `main()`).

WebDrake committed:
Implement blacklist via a CommandLineOption and a validation function

The video driver blacklist can essentially be considered as consisting
of two parts:

  * the `--blacklist` command-line flag, and

  * the init callback, which must be called before the main init
    callback added by `usc::SystemCompositor`.

However, since the `miral::CommandLineOption` class really just adds an
init callback under the hood, we can implement the blacklist directly as
a `CommandLineOption`, whose callback is a new standalone function that
gets the video driver info and performs the blacklist check.

This solution should work well with `miral::MirRunner::run_with`.

WebDrake commented:
Updated with 2 small tweaks:

* the very first patch has been updated to require MirAL >= 1.5, since this is needed for `pre_init` command line options

* an extra patch has been added using a `pre_init` `CommandLineOption` to implement the `--version` flag.

WebDrake commented:
Updated with an extra patch that separates out handling of the video driver blacklist to a separate class, which encapsulates adding both the necessary command-line option and the init callback.

@AlanGriffiths note that in implementing this class I came across an interesting scope-related issue with `CommandLineOption` (https://bugs.launchpad.net/miral/+bug/1712421).  I don't know if it's actually possible to resolve upstream, but my workaround here was to actually store the `CommandLineOption` instance under the hood inside the `Blacklist` class.

AlanGriffiths commented:
@WebDrake:
> Updated with an extra patch that separates out handling of the video driver blacklist to a separate class, which encapsulates adding both the necessary command-line option and the init callback.

This isn't the way I envisaged the MirAL customization tools being used. I realise you are following the existing design, but I think that could be improved too.
I think the blacklist check should be callable (and testable) independently of the server setup. A first step towards that would be making it a function:

```
    void validate_driver(string const& blacklist_regex) { ...}
```
(I realise this is still coupled to the GL context, but it is a step in the right direction.)

Once that is in place you only need CommandLineOption to call it:

```
    CommandLineOption check_blacklist_option{&validate_driver, "blacklist", "Video blacklist regex to use", ""};
```

And I would apply the same design to multi-command options features:

```
    Feature feature; // in a scope with a suitable lifetime

    CommandLineOption opt1{[&feature](string const& val) { feature.opt1(val}; }, ...};
    CommandLineOption opt2{[&feature](string const& val) { feature.opt2(val}; }, ...};
    ...
    CommandLineOption optN{[&feature](string const& val) { feature.optN(val}; }, ...};
    AddInitCallback  apply{[&feature] { feature.apply(); }, ...};
```

WebDrake commented:
@AlanGriffiths OK, I'll try the suggested approach.  Where the blacklist is concerned it's obviously preferable.

Part of the reason for the approach I took was that I wanted to try to abstract away the details of various components where I had a strong suspicion they would be replaceable by corresponding MirAL components in future: it seemed cleaner to abstract everything needed away under the hood (command line options, behaviour, etc.).

But I would agree that probably it makes sense for `main()` to be a single point where all things like custom command line options, etc., are declared.

WebDrake commented:
I've replaced the `Blacklist` class patches with the suggested implementation that just provides a `validate_video_driver(std::string const&)` function, and uses a `CommandLineOption` explicitly in `main()`.

WebDrake committed:
Create display-manager connection via a standalone function

This patch extracts the core logic of how the `DMConnection` instance
is constructed into a separate function, whose input consists of the
needed config parameters.  The function is then called from within the
init callback added by the `SystemCompositor` class, which still (for
now) takes responsibility for extracting the config options from the
underlying server.

This should allow for easy follow-up changes where the logic of these
parts of the callback are extracted into separate `CommandLineOption`
instances applied in `main`.

`mir::optional_value` is used for the to/from file-descriptor parameters
to allow us to easily integrate with `CommandLineOption` in future.  The
default `-1` value used in the `Option::get` calls in `SystemCompositor`
can be ignored, since with `is_set` returning true, it cannot possibly
be used.

WebDrake committed:
Create display manager connection from main using command-line options

This patch reworks the creation of the `DMConnection` instance used in
USC, extracting it from the init callback added under the hood of the
`SystemCompositor` class, and placing it instead in `main`, where it is
instantiated using a collection of `CommandLineOption` instances.  Most
of the latter are implemented using `pre_init` just to ensure that the
resulting values are set correctly before the display manager connection
itself is instantiated; only the `--debug-without-dm` flag is not, since
its callback is the one inside which the instantiation actually happens.
(The use of `pre_init` for the others is probably overkill, but it can't
hurt to make sure that these values are handled as early as possible.)

This allows us to clean up some uses of `add_configuration_option` and
so reduce the number of direct dependencies on the `mir::Server` API.
We can also now remove the explicit `SystemCompositor` constructor, as
the class no longer needs to access the session switcher.

WebDrake commented:
I've pushed a couple of patches that rework the instantiation of the display-manager connection, in line with comments here: https://github.com/yunit-io/unity-system-compositor/issues/1#issuecomment-338431554

This is unfortunately only a partial effort, as dependencies on the `--file` and `--no-file` command-line options mean that it's not entirely possible to avoid use of the `mir::Server::get_options` method: two calls to this are necessary, one in the `CommandLineOption` used to set the `public_socket` flag, the other in an init callback added to extract the name of the socket file.

On the bright side, it's hopefully all progress :-)

WebDrake committed:
Extract socket-file settings using init callback

This patch adds a new `ExtractSocketConfig` class that can in principle
be used with the `MirRunner::run_with` method.  Its constructor requires
a callback accepting two arguments, via which the class will provide the
name of the socket file to use (deriving from the `--file` command-line
flag) and a boolean flag (deriving from `--no-file`) indicating whether
USC should use avoid using a socket file in the first place.

Under the hood, the callback is itself invoked from a pre-init callback,
in order to ensure that the command line arguments will have been parsed
by the time the original callback is invoked.

The new class is used from within the `main` function to avoid needing
to make a direct call to `mir::Server::get_options` when instantiating
the `ExternalSpinner` instance.

WebDrake commented:
I've reworked the recent commits slightly: an extra patch allows for easier extraction of the socket-file settings from the underlying server.  This is a stopgap solution will allow progress in other areas.

WebDrake committed:
Implement --spinner flag using a pre-init CommandLineOption

Following earlier changes to how the socket file name is obtained, this
patch completes the process of removing any direct dependency on the
`mir::Server` API when constructing the `ExternalSpinner` instance.  A
`pre_init` command-line option is used to ensure that the `spinner_path`
variable will be set as early as possible in the app lifecycle.

The instantiation itself of course still occurs within a callback that
is passed directly to the server API, but this will be dealt with at a
later time.

WebDrake committed:
Implement --public-socket flag using a pre-init CommandLineOption

This removes one more direct dependency on `add_configuration_option`.
Note that it does not simply pass through whether or not the flag is
set, but combines it with the settings for the `--no-file` flag: this
can be compared to how the `--public-socket` and `--no-file` flags are
combined when instantiating the display-manager connection.

This change is not immediately used within the application, but will be
helpful when extracting the creation of the display-manager connection
from within `SystemCompositor`.  `pre_init` is used as a precaution in
order to guarantee that the settings from this flag will be available
for use as early as possible.

WebDrake commented:
Further reworking: I've found a more `MirRunner`-compatible way to extract the socket config, and rebased the other patches on top of that.

WebDrake committed:
Implement --enable-hardware-cursor flag using pre-init CommandLineOption

This patch eliminates the last remaining `add_configuration_option` call
from within the `main` function.  Note that this needs to be implemented
as a `pre_init` command-line option in order to ensure that it will be
invoked before the callback that uses the actual flag value.

WebDrake commented:
I've pushed some further patches to clear up the use of `add_configuration_option`.  However, this has shown up a few potentially nasty failure-cases in terms of how some of the callbacks will be ordered.  Using `pre_init` for the `--enable-hardware-cursor` flag suffices (as long as the flag set is accessed by ref in the later callbacks that use it), but I need to look into how the other settings will interact.

I'm also minded to rewrite all the delegate captures to use `[&]` instead of trying to be precise with the chosen variables.  The latter is not necessary, and runs the risk of bugs like the one I ran into here when using `[enable_hardware_cursor]` instead of `[&enable_hardware_cursor]` in the callback provided to the `wrap_cursor_listener` method.  It'll make things easier for any later developer if callbacks can just read and write anything in scope.

WebDrake commented on src/extract_socket_config.cpp:
```diff
@@ -0,0 +1,39 @@
+
+void usc::ExtractSocketConfig::operator()(mir::Server& server) const
+{
+    server.add_init_callback([&]
```
Wondering whether this should maybe be `add_pre_init_callback` to ensure that it is invoked before any possible downstream user of the flags' values.

WebDrake commented on src/main.cpp:
```diff
@@ -17,19 +17,329 @@
+    };
+    extract_socket_config(*config);
+
+    miral::CommandLineOption public_socket_flag{
```
This should probably be `pre_init` just to be on the safe side

WebDrake commented on src/main.cpp:
```diff
@@ -17,19 +17,329 @@
+
+    // spinner settings
+    std::string spinner_path;
+    miral::CommandLineOption spinner_flag{
```
Should probably be `pre_init` just to be sure that it is available by the time it's used (may be worth checking this in any case, similar to what's been done for `--enable-hardware-cursor`).

WebDrake commented:
In line with recent observations I've reworked all the patches from the removal of `usc::Server` on, in order to:

* replace specific capture rules for lambdas with `[&]` (which makes for a simpler set of assumptions, cleaner patches, and avoids nasty details about which variables are accessed by value or by reference)

* switch `CommandLineOption` instances to use `pre_init` where appropriate (i.e. for all command-line options that are just setting config values and not actually _doing_ anything).

These changes among other things fix a bug with the `CommandLineOption` implementation of the `--spinner` flag, similar to that already observed with the original `--enable-hardware-cursor` changes.

WebDrake commented:
OK, it looks like there's a problem in the `Create display manager connection from main ...` patch.  USC obviously fails to start in some deeply problematic way following this change.  I'm a bit too tired to start debugging it right now, so will come back to it some time in the next days.

WebDrake commented on src/main.cpp:
```diff
@@ -238,11 +233,70 @@ try
+    std::shared_ptr<usc::DMConnection> dm_connection;
+
+    miral::CommandLineOption debug_without_dm_flag{
+        [&](bool debug_without_dm)
```
At a guess, this is only being invoked if the flag is actually set -- which has always been the case in my local tests, but stopped being true once I took it out for a spin on my Unity8 system.  So it might be a good idea to only set a flag using the `CommandLineOption`, and use an explicit `AddInitCallback` to initialize `dm_connection`.

WebDrake replied:
The remark here seems to be completely off base.  Current version of code works without needing to change this particular `CommandLineOption`.

WebDrake commented:
Using an explicit `AddInitCallback` (as in the last `squash!` patch) doesn't seem to help.  I'll come back to this tomorrow :-P

WebDrake commented:
Reworked both the `create_dm_connection` function and the fd-related `CommandLineOption` declarations using `mir::optional_value`.  This allows reconstruction of the original app logic.    @AlanGriffiths thanks for help and advice on these points!

It's difficult to really see why, but for whatever reason this fixes the problem introduced in the earlier version of the `Create display manager connection from main ...` patch.  So it looks like I can continue building on top of this.

WebDrake commented on src/main.cpp:
```diff
@@ -238,11 +233,76 @@ try
+        [&](mir::optional_value<int> const& to_dm_file_descriptor)
+        {
+            to_dm_fd = to_dm_file_descriptor;
+        };
```
Note: seems to be necessary to instantiate these callbacks as lvalues in order for the `CommandLineOption` constructor to accept them.  Otherwise it seems to create some ambiguity when the compiler tries to resolve which overload of the constructor to choose.

WebDrake committed:
Instantiate session_switcher in --spinner CommandLineOption callback

This patch ensures that the `spinner` and `session_switcher` can be set
using only the MirAL API.  Since `pre_init` is used, `session_switcher`
should be instantiated before the `override_the_window_manager_builder`
callback is invoked.  The latter callback is now limited to doing only
the `usc::WindowManager` instantiation that it needs to do.

WebDrake commented:
I've pushed one further patch, a refinement of the `--spinner` flag `CommandLineOption` that takes care of instantiating the `session_switcher` variable.

This leaves us with a few reasonably concentrated entities that have a dependency on the mirserver API.

Inside `main.cpp`, then the code concerned is a series of calls to various wraps and overrides:

```C++
config->set_command_line(argc, const_cast<char const **>(argv));
config->set_command_line_handler(&ignore_unknown_arguments);
config->wrap_cursor_listener(...);
config->override_the_server_status_listener(...);
config->override_the_window_manager_builder(...);
config->override_the_cookie_authority(...);
config->set_config_filename("unity-system-compositor.conf");
```

Of these, the only one with any meaningful overlap with the rest of `main.cpp` is the `override_the_window_manager_builder` call, which uses `session_switcher` inside the callback passed to it.

Over in `system_compositor.cpp` we have the variety of stuff that is still added manually via `mir::Server::add_init_callback`, some of which has heavy mirserver dependencies.

First, we have the instantiation of the `MirScreen` instance.  `MirScreen` is a subclass of the `usc::Screen` base class (which AFAICS has no mir dependencies), and `mir::graphics::DisplayConfigurationObserver` (which obviously is a heavy mir dependency).

Once instantiated, this is registered directly with the server display configuration observer registrar (mirserver dependency), and then, via the `Screen` interface, is used to instantiate the `UnityDisplayService` instance.  The latter has no mir dependencies (it's all based around dbus).

The next few creations (`UnityPowerButtonEventSink`, `UnityUserActivityEventSink`, `SteadyClock`) have no mir dependency, but  `ScreenEventHandler` (which is constructed using these) is a subclass of `mir::input::EventFilter` (and uses various `mir::time` things under the hood).

The `ScreenEventHandler` instance is registered with the server via `mir::Server::the_composite_event_filter()->append`.

We next instantiate a `MirInputConfiguration`.  This takes a `mir::input::InputDeviceHub` as a constructor parameter, and uses other `mir::input` under the hood.

This is used to instantiate a `UnityInputService` instance (which doesn't have any direct mir dependencies, it's all based around dbus).

Finally we have the construction of a `DBusConnectionThread`, instantiated using a `DBusEventLoop` created earlier in the callback: neither of these has any relationship with any mir API.
